packages/block-editor/src/components/block-edit/index.js

```diff
@@ -1,52 +1,17 @@
-/**
- * WordPress dependencies
- */
-import { Component } from '@wordpress/element';
-
 /**
  * Internal dependencies
  */
 import Edit from './edit';
 import { BlockEditContextProvider } from './context';
 
-class BlockEdit extends Component {
-	constructor( props ) {
-		super( props );
-
-		this.setFocusedElement = this.setFocusedElement.bind( this );
-
-		this.state = {
-			focusedElement: null,
-			setFocusedElement: this.setFocusedElement,
-		};
-	}
-
-	setFocusedElement( focusedElement ) {
-		this.setState( ( prevState ) => {
-			if ( prevState.focusedElement === focusedElement ) {
-				return null;
-			}
-			return { focusedElement };
-		} );
-	}
-
-	static getDerivedStateFromProps( props ) {
-		const { clientId, name, isSelected } = props;
-
-		return {
-			name,
-			isSelected,
-			clientId,
-		};
-	}
+function BlockEdit( props ) {
+	const { name, isSelected, clientId } = props;
 
-	render() {
-		return (
-			<BlockEditContextProvider value={ this.state }>
-				<Edit { ...this.props } />
-			</BlockEditContextProvider>
-		);
-	}
+	return (
+		<BlockEditContextProvider value={ { name, isSelected, clientId } }>
```

Comment:
It must go through state management otherwise will trigger re-render to every `BlockEdit` tree each time anything changes in props because new object reference is going to be passed as `value`.

Reply:
See: https://reactjs.org/docs/context.html#caveats

> Because context uses reference identity to determine when to re-render, there are some gotchas that could trigger unintentional renders in consumers when a provider’s parent re-renders.

Reply:
Thanks. It sounds like we just need to ensure we keep the same reference if the props don't change, it must not necessarily be dependent on the state? I'll have a closer look.

Reply:
Yes, React team recommends using state in their docs but you can find a different solution as well.

Reply:
Hm, they recommend state, but in `master`, we're also not setting the state properly, which completely defeats the purpose of using the state to keep the same reference.

> `getDerivedStateFromProps` is invoked right before calling the render method, both on the initial mount and on subsequent updates. It should return an object to update the state, or null to update nothing.
> https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops

See also https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#anti-pattern-unconditionally-copying-props-to-state.

We're always passing a new object on every render:

https://github.com/WordPress/gutenberg/blob/a61e351234d9c74f9fe80b35f6f97769e0758d78/packages/block-editor/src/components/block-edit/index.js#L33-L41

It's probably better to use memoization instead, as a recommended alternative to `getDerivedStateFromProps`:

> If you want to re-compute some data only when a prop changes, use a memoization helper instead.
> https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops

Reply:
Fixed in 5e362950d.

Reply:
See this comment from @aduth: https://github.com/WordPress/gutenberg/pull/7192#discussion_r193554592. In his testing, it worked as a patch operation and it didn't cause additional re-renders. 

I don't know if using memoization with history level set to 1 will work here as it needs to keep local state for each block rendered in the editor.

Reply:
Oh, right. It needs to be bound to the class instance instead.

Reply:
Fixed in 69eb5f5fd702fda101b691d467eed6089c6e0888 (moved into a class).